utils/formatting/encoding.go

```diff
@@ -4,51 +4,141 @@
 package formatting
 
 import (
+	"bytes"
+	"encoding/hex"
+	"errors"
 	"fmt"
+	"math"
+	"strings"
+
+	"github.com/ava-labs/avalanchego/utils/hashing"
+	"github.com/mr-tron/base58/base58"
 )
 
 const (
-	// HexEncoding specifies a hex plus 4 byte checksum encoding format
-	HexEncoding = "hex"
-	// CB58Encoding specifies the CB58 encoding format
-	CB58Encoding = "cb58"
+	// maximum length byte slice can be marshalled to a string
+	// using the CB58 encoding. Must be longer than the length
+	// of an ID and longer than the length of a SECP256k1 private key
+	// TODO: Reduce to a reasonable amount (e.g. 16 * 1024) after we
+	// give users a chance to export very large keystore users to hex
+	maxCB58Size = math.MaxInt32
+
+	checksumLen = 4
+)
+
+var (
+	errInvalidEncoding  = errors.New("invalid encoding")
+	errMissingChecksum  = errors.New("input string is smaller than the checksum size")
+	errBadChecksum      = errors.New("invalid input checksum")
+	errMissingHexPrefix = errors.New("missing 0x prefix to hex encoding")
 )
 
-// Encoding returns a struct used to format bytes for a specific encoding
-type Encoding interface {
-	ConvertBytes([]byte) string
-	ConvertString(str string) ([]byte, error)
-	Encoding() string
+// Encoding defines how bytes are converted to a string and vice versa
+type Encoding uint8
+
+const (
+	// Hex specifies a hex plus 4 byte checksum encoding format
+	Hex Encoding = iota
+	// CB58 specifies the CB58 encoding format
+	CB58
+)
+
+// String ...
+func (enc Encoding) String() string {
+	switch enc {
+	case Hex:
+		return "hex"
+	case CB58:
+		return "cb58"
+	default:
+		return errInvalidEncoding.Error()
+	}
 }
 
-// EncodingManager is an interface to provide an Encoding interface
-type EncodingManager interface {
-	GetEncoding(encoding string) (Encoding, error)
+func (enc Encoding) valid() bool {
+	switch enc {
+	case Hex, CB58:
+		return true
+	}
+	return false
 }
 
-type manager struct {
-	defaultEnc string
+// MarshalJSON ...
+func (enc Encoding) MarshalJSON() ([]byte, error) {
+	if !enc.valid() {
+		return nil, errInvalidEncoding
+	}
+	return []byte("\"" + enc.String() + "\""), nil
 }
 
-// NewEncodingManager returns an EncodingManager with the provided default
-func NewEncodingManager(defaultEnc string) (EncodingManager, error) {
-	if defaultEnc != HexEncoding && defaultEnc != CB58Encoding {
-		return nil, fmt.Errorf("unrecognized default encoding: %s", defaultEnc)
+// UnmarshalJSON ...
+func (enc *Encoding) UnmarshalJSON(b []byte) error {
```

Comment:
The json interface specifies that the implementing struct should be able to take in `null` and it should be equal to not having called `Unmarshal`.